src/app/api/game-sessions/route.ts

```diff
@@ -1,40 +1,371 @@
 
-    const body = insertGameSessionSchema.parse(json);
+    const gameSessionToInsert = insertGameSessionSchema.parse(await req.json());
+    const gameSessionDate = gameSessionToInsert.date;
 
-    if (new Date(body.date) < new Date()) {
+    if (gameSessionDate < formatInNZST(new Date())) {
       return new Response("Date cannot be in the past", { status: 400 });
     }
 
-    await db.insert(gameSessions).values(body);
+    // Check the game session is valid
+    if (formatInNZST(gameSessionToInsert.bookingOpen) > gameSessionDate) {
+      return new Response("Booking open must be before the game session date", {
+        status: 400,
+      });
+    }
 
-    return NextResponse.json(body, { status: 201 });
+    if (formatInNZST(gameSessionToInsert.bookingClose) > gameSessionDate) {
+      return new Response(
+        "Booking close must be before the game session date",
+        {
+          status: 400,
+        },
+      );
+    }
+
+    if (gameSessionToInsert.bookingOpen > gameSessionToInsert.bookingClose) {
+      return new Response("Booking open must be before booking close", {
+        status: 400,
+      });
+    }
+
+    if (gameSessionToInsert.startTime > gameSessionToInsert.endTime) {
+      return new Response("Start time must be before end time", {
+        status: 400,
+      });
+    }
+
+    if (gameSessionToInsert.casualCapacity > gameSessionToInsert.capacity) {
+      return new Response("Casual capacity must be less than capacity", {
+        status: 400,
+      });
+    }
+
+    // Check that a record doesn't already exist for this date
+    const existingGameSession = await db.query.gameSessions.findFirst({
+      where: eq(gameSessions.date, gameSessionDate),
+    });
+
+    if (existingGameSession) {
+      return new Response("A game session already exists for this date", {
+        status: 400,
+      });
+    }
+
+    // Check for a schedule on this weekday
+    const semester = await db.query.semesters.findFirst({
+      with: {
+        gameSessionSchedules: {
+          where: eq(gameSessionSchedules.weekday, getWeekday(gameSessionDate)),
+        },
+      },
+      where: or(
+        and(
+          lte(semesters.startDate, gameSessionDate),
+          gt(semesters.breakStart, gameSessionDate),
+        ),
+        and(
+          lt(semesters.breakEnd, gameSessionDate),
+          gte(semesters.endDate, gameSessionDate),
+        ),
+      ),
+    });
+
+    const gameSessionSchedule = semester?.gameSessionSchedules[0];
+
+    if (gameSessionSchedule) {
+      // If a schedule exists, only create record if an exception also exists, otherwise it will be generated
+      const gameSessionException =
+        await db.query.gameSessionExceptions.findFirst({
+          where: eq(gameSessionExceptions.date, gameSessionDate),
+        });
+
+      if (!gameSessionException) {
+        return new Response("No gameSessionException found for this date", {
+          status: 400,
+        });
+      }
+    }
+
```

Comment:
For POSTs we want to insert an exception ONLY IF
* game-session doesn't exist (already checked)
* gameSessionSchedule doesn't exist
* gameSessionSchedule exists AND gameSessionException.isDeleted = true

you might have to edit this check to account for the last two.`